llvm/lib/Transforms/IPO/AttributorAttributes.cpp

```diff
@@ -12571,17 +12571,59 @@ struct AAAddressSpaceImpl : public AAAddressSpace {
   void initialize(Attributor &A) override {
     assert(getAssociatedType()->isPtrOrPtrVectorTy() &&
            "Associated value is not a pointer");
-    if (getAssociatedType()->getPointerAddressSpace())
+    // If the pointer already has non-generic address space, we assume it is the
+    // correct one.
+    if (getAssociatedType()->getPointerAddressSpace()) {
+      [[maybe_unused]] bool R =
+          takeAddressSpace(getAssociatedType()->getPointerAddressSpace());
+      assert(R && "the take should happen");
       indicateOptimisticFixpoint();
+      return;
+    }
+    // If the pointer is an addrspacecast, we assume the source address space is
+    // the correct one.
+    Value *V = &getAssociatedValue();
+    if (auto *ASC = dyn_cast<AddrSpaceCastInst>(V)) {
+      [[maybe_unused]] bool R = takeAddressSpace(ASC->getSrcAddressSpace());
+      assert(R && "the take should happen");
+      indicateOptimisticFixpoint();
+      return;
+    }
+    if (auto *C = dyn_cast<ConstantExpr>(V)) {
+      if (C->getOpcode() == Instruction::AddrSpaceCast) {
+        [[maybe_unused]] bool R = takeAddressSpace(
+            C->getOperand(0)->getType()->getPointerAddressSpace());
+        assert(R && "the take should happen");
+        indicateOptimisticFixpoint();
+        return;
+      }
+    }
   }
 
   ChangeStatus updateImpl(Attributor &A) override {
-    int32_t OldAddressSpace = AssumedAddressSpace;
+    uint32_t OldAddressSpace = AssumedAddressSpace;
     auto *AUO = A.getOrCreateAAFor<AAUnderlyingObjects>(getIRPosition(), this,
                                                         DepClassTy::REQUIRED);
     auto Pred = [&](Value &Obj) {
       if (isa<UndefValue>(&Obj))
         return true;
+      // If an argument in generic address space has addrspace cast uses, and
+      // those casts are same, then we take the dst addrspace.
+      if (auto *Arg = dyn_cast<Argument>(&Obj)) {
+        if (Arg->getType()->getPointerAddressSpace() == 0) {
```

Comment:
should query the target generic address space and not assume 0

Reply:
Well, I understand I should use `TargetTransformInfo::getFlatAddressSpace`, but the fact is it doesn't always work. It requires to pass a function to query it.  Here, the associated value can be a `Constant` which doesn't associate with any function, therefore I can't get TTI. It's quite unfortunate that the flat address space information is part of a function analysis result. Is there any way to work around it? I have an ugly solution, which is to pick a function from the module randomly, query the flat address space, and store it in the information cache. I don't really want to do that.

Reply:
getFlatAddressSpace really belongs in TargetMachine, not TTI

Reply:
I posted a [PR](https://github.com/llvm/llvm-project/pull/108594) to add it in `TargetMachine`, then I run into another issue: `TargetMachine` is also not available here. Unlike `AMDGPUAttributor`, AAs in `llvm/lib/Transforms/IPO/AttributorAttributes.cpp` as well as their user passes, may not have access to `TargetMachine`.